spec/cancan/model_adapters/active_record_adapter_spec.rb

```diff
@@ -237,101 +237,84 @@ class User < ActiveRecord::Base
       @ability.can :read, Article
       @ability.cannot :read, Article, published: false, secret: true
-      expect(@ability.model_adapter(Article, :read).conditions)
-        .to orderlessly_match(%["not (#{@article_table}"."published" = 'f' AND "#{@article_table}"."secret" = 't')])
+      expect(@ability.model_adapter(Article, :read).database_records.to_sql.strip.squeeze(' '))
+        .to eq(%q(SELECT DISTINCT articles.* FROM (SELECT "articles".* FROM "articles" EXCEPT SELECT "articles".* FROM "articles" WHERE "articles"."published" = 'f' AND "articles"."secret" = 't') AS articles))
     end
 
     it 'returns appropriate sql conditions in complex case' do
       @ability.can :read, Article
       @ability.can :manage, Article, id: 1
       @ability.can :update, Article, published: true
       @ability.cannot :update, Article, secret: true
-      expect(@ability.model_adapter(Article, :update).conditions)
-        .to eq(%[not ("#{@article_table}"."secret" = 't') ] +
-               %[AND (("#{@article_table}"."published" = 't') ] +
-               %[OR ("#{@article_table}"."id" = 1))])
-      expect(@ability.model_adapter(Article, :manage).conditions).to eq(id: 1)
-      expect(@ability.model_adapter(Article, :read).conditions).to eq("'t'='t'")
+      expect(@ability.model_adapter(Article, :update).database_records.to_sql.strip.squeeze(' '))
+        .to eq(%q(SELECT DISTINCT articles.* FROM (SELECT "articles".* FROM "articles" WHERE "articles"."id" = 1 UNION SELECT "articles".* FROM "articles" WHERE "articles"."published" = 't' EXCEPT SELECT "articles".* FROM "articles" WHERE "articles"."secret" = 't') AS articles))
+      expect(@ability.model_adapter(Article, :manage).database_records.to_sql.strip.squeeze(' '))
+        .to eq(%q(SELECT DISTINCT articles.* FROM (SELECT "articles".* FROM "articles" WHERE "articles"."id" = 1) AS articles))
+      expect(@ability.model_adapter(Article, :read).database_records.to_sql.strip.squeeze(' '))
+        .to eq(%q(SELECT DISTINCT articles.* FROM (SELECT "articles".* FROM "articles" UNION SELECT "articles".* FROM "articles" WHERE "articles"."id" = 1) AS articles))
+    end
+
+    it 'returns appropriate sql conditions in cases where two rules specify different conditions on the same remote table, accessed via distinct joins' do
+      @ability.can :read, Article, user: { id: 1 }
+      @ability.can :read, Article, { mentioned_users: { id: 2 } }
+      expect(@ability.model_adapter(Article, :read).database_records.to_sql.strip.squeeze(' '))
+        .to eq(%q(SELECT DISTINCT articles.* FROM (SELECT "articles".* FROM "articles" INNER JOIN "users" ON "users"."id" = "articles"."user_id" WHERE "users"."id" = 1 UNION SELECT "articles".* FROM "articles" INNER JOIN "legacy_mentions" ON "legacy_mentions"."article_id" = "articles"."id" INNER JOIN "users" ON "users"."id" = "legacy_mentions"."user_id" WHERE "users"."id" = 2) AS articles))
```

Comment:
Nope - UNIONs are at the heart of what this PR enables. Recall from the PR notes that previously cancancan would split off the logic that generates joins from the condition logic quite early in the query generation process, making it impossible to produce queries on a per-join basis. The heart of this PR is inverting that behaviour and instead producing subqueries for each ability with the requisite joins and conditions being self-contained. These subqueries are then joined via SQL UNION statements to produce the desired net effect